mlflow/pipelines/steps/automl/flaml.py

```
@@ -0,0 +1,124 @@
+    # pylint: disable=unused-argument
+    def custom_metric(
+        X_val,
+        y_val,
+        estimator,
+        labels,
+        X_train,
+        y_train,
+        weight_val=None,
+        weight_train=None,
+        *args,
+    ):
+        val_metrics = calc_metric(X_val, y_val, estimator)
+        if metric_name not in val_metrics:
+            raise MlflowException(
+                f"User-defined function has not calculated expected primary metric {metric_name}.\n"
+                f"The function has returned the following metrics: {val_metrics}"
+            )
+        main_metric = coeff * val_metrics[metric_name]
+        val_metrics = add_suffix(val_metrics, "val")
+        train_metrics = add_suffix(calc_metric(X_train, y_train, estimator), "train")
+        return main_metric, {**val_metrics, **train_metrics}
+
+    return custom_metric
+
+
+def _create_model_automl(
+    X,
+    y,
+    step_config: Dict[str, Any],
+    pipeline_root: str,
+    evaluation_metrics: Dict[str, PipelineMetric],
+    primary_metric: str,
+) -> Tuple[BaseEstimator, Dict[str, Any]]:
+    try:
+        from flaml import AutoML
+    except ImportError:
+        raise MlflowException("Please install FLAML to use AutoML!")
+
+    try:
+        if primary_metric in MLFLOW_TO_FLAML_METRICS and primary_metric in evaluation_metrics:
+            metric = MLFLOW_TO_FLAML_METRICS[primary_metric]
+        elif primary_metric in evaluation_metrics:
+            metric = _create_custom_metric_flaml(
+                primary_metric,
+                -1 if evaluation_metrics[primary_metric].greater_is_better else 1,
+                _load_one_custom_metric_function(pipeline_root, evaluation_metrics[primary_metric]),
+            )
+        else:
+            _logger.warning(
+                f"There is no FLAML alternative or custom metric for {primary_metric} metric.\n"
+                f"Using 'auto' metric instead."
+            )
+            metric = "auto"
```

I think that we should throw an error in this case. We should require the user to specify a primary metric (which we do), and either respect that metric or error.

---

done